sdk/core/Azure.Core/src/Shared/DiagnosticScope.cs

```diff
@@ -370,9 +378,56 @@ static ActivityExtensions()
         private static Func<object, bool>? ActivitySourceHasListenersMethod;
         private static Func<string, string?, ICollection<KeyValuePair<string, object>>?, object?>? CreateActivityLinkMethod;
         private static Func<ICollection<KeyValuePair<string,object>>?>? CreateTagsCollectionMethod;
-
+        private static Func<Activity, string, object?>? GetCustomPropertyMethod;
+        private static Action<Activity, string, object>? SetCustomPropertyMethod;
+        private static string AzureSdkScopeLabel = "az.sdk.scope";
         private static readonly ParameterExpression ActivityParameter = Expression.Parameter(typeof(Activity));
 
+        public static bool IsNestedScopeEnabled(this Activity activity)
+        {
+            if (GetCustomPropertyMethod == null)
```

Comment:
unfortunately, we can't use our own `AsyncLocal`. the problem is that we use shared sources in clietn libraries and have multiple copies of `ClientDiagnsotics`, `DiagnsoticScope`, etc classes. So we can't rely on shared static properties.

So we have to rely on something public, which is `Activity`. 
We should consider making DiagnosticScope public and then we can make these methods more efficient.

Reply:
Would it be possible to use the [Baggage ](https://docs.microsoft.com/en-us/dotnet/api/system.diagnostics.activity.setbaggage?view=net-6.0)collection on the Activity itself to stash the context we need?

Reply:
Good question, Baggage is something that propagates between services, there is no way to remove the baggage (ever), and is intended for applications to use.

Reply:
According to [SetBaggage](https://docs.microsoft.com/en-us/dotnet/api/system.diagnostics.activity.setbaggage?view=net-6.0) docs, items can be removed:

> Add or update the Activity baggage with the input key and value. If the input value is null - if the collection has any baggage with the same key, then this baggage will get removed from the collection.

It seemed like it could be appropriate because this is an application specific use case.

Reply:
Ah, I didn't know you could remove it. Thanks! Still, it's not the right place to put out flags into since what will happen is:
- we start Activity with Baggage
- all its child activities will have copies of this package - in case user callback is called anywhere - there can be many of them which we have no control over
- we'll need to remove the baggage before headers are set - Baggage goes into headers
- we'll need to add baggage again after request is done to suppress anything under.

So, Baggage is not designed for it, CustomProperties are designed exactly for it.